sdk/lib/opentelemetry/sdk/trace/samplers/consistent_probability_tracestate.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Trace
+      module Samplers
+        # @api private
+        #
+        # The ConsistentProbabilityTraceState module implements common TraceState
+        # validation and manipulation for the consistent probability-based samplers.
+        module ConsistentProbabilityTraceState
+          DECIMAL = /\A\d+\z/.freeze
+          private_constant(:DECIMAL)
+
+          def validate_tracestate(span_context) # rubocop:disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+            sampled = span_context.trace_flags.sampled?
+            tracestate = span_context.tracestate
+            ot = OpenTelemetry::SDK::Trace::Tracestate.from_tracestate(tracestate)
+            r = decimal(ot.value('r'))
+            p = decimal(ot.value('p'))
+            new_ot = ot
+            new_ot = ot.delete('p') if !p.nil? && p > 63
+            new_ot = ot.delete('p').delete('r') if !r.nil? && r > 62
+            r = nil if !r.nil? && r > 62 # TODO: clean up this logic
+            new_ot = ot.delete('p') if !p.nil? && !r.nil? && !invariant(p, r, sampled)
```

Comment:
It is hard to follow. I explicitly asked Robert for his opinion on this, since I found it messy, but it captures the requirements and we had a hard time finding a clearer way to express it in code. Suggestions welcome.

Reply:
I have a substantial refactoring (after reading the Go implementation) that might clean this up a bit. I'll try to push that up today.